Fix filename completions on a nonexisting file outside your home directory

ujihisa opened the pull request:
## Problem

ddc-file does not provide filenames when all of the following conditions have met:

* vim8
* not at a normal file (e.g. deol-edit buffer)
* not at your home directory

https://asciinema.org/a/rIvUFnqEAJhzm5stMSLu1KUUM

## Solution

This patch is not beautiful at all but it solves the problem.

https://asciinema.org/a/J3F5dGk85AOY8m6rxN5A1cVJE

Shougo commented on autoload/ddc_file/internal.vim:
```diff
@@ -26,7 +26,13 @@ endfunction
   let input_file_base_prefix = s:full_to_base_prefix(input_file_full, a:is_posix)
-  let buf_path = expand('%:p')
+
+  if &buftype ==# '' && expand('%:p') !=# ''
```
I think it should be `if &buftype ==# '' && !isdirectory(expand('%:p:h'))`

Shougo commented:
`buf_path` is used to detect buffer directory.  So `buf_path` is invalid, you cannot complete path.

ujihisa commented:
Added 7bc210550eef778edbd20c6730b046f668e06610 to fix the known issue described in the description, and updated the description itself!

Removed part:
 
> ## Known issue
> 
> It keeps the files at the previous directory too...

Shougo commented:
```diff
diff --git a/denops/@ddc-sources/file.ts b/denops/@ddc-sources/file.ts
index 47efd4e..c55e819 100644
--- a/denops/@ddc-sources/file.ts
+++ b/denops/@ddc-sources/file.ts
@@ -2,6 +2,7 @@ import {
   BaseSource,
   Candidate,
   GatherCandidatesArguments,
+  fn,
   homeDir,
   path as univPath,
   wrapA,
@@ -139,7 +140,7 @@ export class Source extends BaseSource<Params> {

     // point from cwd
     findPointsAsync.push({
-      dir: Deno.cwd(),
+      dir: await fn.getcwd(args.denops) as string,
       max: p.cwdMaxCandidates,
       menu: p.displayCwd,
       asRoot: p.cwdAsRoot,
```

This is better.

ujihisa commented:
@Shougo Thx! Added 289fc133e3c7e9d004f60c8bd2c38b950825fc26 to use its getcwd.

ujihisa commented:
Oops

LumaKernel commented:
Thank you! Formatting looks failing. Can you run `deno fmt`?

LumaKernel commented:
https://github.com/LumaKernel/ddc-file/blob/6ea53360935eceecd2882389a70c0381bdbaabdc/denops/%40ddc-sources/file.ts#L152

Could you please fix here too with storing evaluated cwd into variable?

ujihisa commented:
Thx! Fixed both of them.

LumaKernel commented:
Merged. Much appreciated!

ujihisa commented:
Thx!
